plugins/meta/route-override/route-override.go

```diff
@@ -0,0 +1,400 @@
+	}
+
+	return err
+}
+
+func addRoute(dev netlink.Link, route *types.Route) error {
+	return netlink.RouteAdd(&netlink.Route{
+		LinkIndex: dev.Attrs().Index,
+		Scope:     netlink.SCOPE_UNIVERSE,
+		Dst:       &route.Dst,
+		Gw:        route.GW,
+	})
+}
+
+func processRoutes(netnsname string, conf *RouteOverrideConfig) (*current.Result, error) {
+	netns, err := ns.GetNS(netnsname)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open netns %q: %v", netns, err)
+	}
+	defer netns.Close()
+
+	res, err := current.NewResultFromResult(conf.PrevResult)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert result to current version: %v", err)
+	}
+
+	if conf.FlushGateway {
+		// add "0.0.0.0/0" into delRoute to remove it from routing table/result
+		_, gwRoute, _ := net.ParseCIDR("0.0.0.0/0")
+		conf.DelRoutes = append(conf.DelRoutes, &types.Route{Dst: *gwRoute})
+		_, gwRoute, _ = net.ParseCIDR("::/0")
+		conf.DelRoutes = append(conf.DelRoutes, &types.Route{Dst: *gwRoute})
+
+		// delete given gateway address
+		for _, ips := range res.IPs {
+			if ips.Version == "6" {
+				ips.Gateway = net.IPv6zero
+			} else {
+				ips.Gateway = net.IPv4zero
+			}
+		}
+	}
+
+	newRoutes := []*types.Route{}
+	err = netns.Do(func(_ ns.NetNS) error {
+		// Flush route if required
+		if !conf.FlushRoutes {
+		NEXT:
+			for _, route := range res.Routes {
+				for _, delroute := range conf.DelRoutes {
+					if route.Dst.IP.Equal(delroute.Dst.IP) &&
+						bytes.Equal(route.Dst.Mask, delroute.Dst.Mask) {
+						err = deleteRoute(delroute, res)
+						if err != nil {
+							fmt.Fprintf(os.Stderr, "failed to delte route %v: %v", delroute, err)
+						}
+						continue NEXT
+					}
+
+				}
+				newRoutes = append(newRoutes, route)
+			}
+		} else {
+			deleteAllRoutes(res)
+		}
+
+		if conf.FlushGateway {
+			deleteGWRoute(res)
+		}
+
+		// Get container IF name
+		var containerIFName string
+		for _, i := range res.Interfaces {
+			if i.Sandbox != "" {
+				containerIFName = i.Name
+				break
+			}
+		}
+		// Add route
+		dev, _ := netlink.LinkByName(containerIFName)
+		for _, route := range conf.AddRoutes {
+			newRoutes = append(newRoutes, route)
+			if err := addRoute(dev, route); err != nil {
+				fmt.Fprintf(os.Stderr, "failed to add route: %v: %v", route, err)
+			}
+		}
+
+		return nil
+	})
+	res.Routes = newRoutes
+
+	return res, nil
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	overrideConf, err := parseConf(args.StdinData, args.Args)
+	if err != nil {
+		return err
+	}
+
+	newResult, err := processRoutes(args.Netns, overrideConf)
+	if err != nil {
+		return fmt.Errorf("failed to override routes: %v", err)
+	}
+
+	return types.PrintResult(newResult, overrideConf.CNIVersion)
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+	// TODO: the settings are not reverted to the previous values. Reverting the
+	// settings is not useful when the whole container goes away but it could be
+	// useful in scenarios where plugins are added and removed at runtime.
```

Comment:
This can be a really big issue as we can not expect that next plugin in the chain will just ignore that configuration is different from what it left after it was called for `ADD` command.
There can be (i'm not sure but probably calico is an example) plugins "in the wild" which will find that undesirable.

Reply:
I'm not clear about it. 

As far as I checked, [calico CNI plugin](https://github.com/projectcalico/cni-plugin) does not touch routes in cmdDel() as following:

https://github.com/projectcalico/cni-plugin/blob/master/internal/pkg/utils/network_linux.go#L437-L478

Reply:
It was only an example (as You pointed - missed one).
Also afair multus can work (or is planned to work) in "thick plugin" mode which will allow dynamic calling add/remove on plugins - in which case this will also lead to issues.

Reply:
Currently from multus side, none working for "thick plugin" because of less usecase/requirements, so I suppose your concern is valid but target to future. How about keep it in TODO (or issue) to track it?